apps/extension/src/state/tx-approval.ts

```diff
@@ -13,15 +23,114 @@ export interface TxApprovalSlice {
   transactionClassification?: TransactionClassification;
-  // Holding the message responder function. Service worker will be "awaiting" the call of this.
-  responder?: MessageResponder<TxApproval>;
+
+  attitude?: boolean;
+  setAttitude: (attitude: boolean) => void;
+
+  sendResponse: () => void;
 }
 
-export const createTxApprovalSlice: SliceCreator<TxApprovalSlice> = () => ({});
+export const createTxApprovalSlice = (): SliceCreator<TxApprovalSlice> => (set, get) => ({
+  acceptRequest: async (
+    { authorizeRequest: authReqJson, transactionView: txViewJson },
+    responder,
+  ) => {
+    const authorizeRequest = AuthorizeRequest.fromJson(authReqJson);
+    const transactionView = TransactionView.fromJson(txViewJson);
+
+    const asSender = transactionView;
+    const asPublic = asPublicTransactionView(transactionView);
+    const asReceiver = await asReceiverTransactionView(transactionView, {
+      isControlledAddress,
+    });
+
+    set(state => {
+      state.txApproval.responder = responder;
+      state.txApproval.authorizeRequest = authorizeRequest.toJsonString();
+      state.txApproval.transactionView = transactionView.toJsonString();
+
+      state.txApproval.asSender = asSender.toJsonString();
+      state.txApproval.asPublic = asPublic.toJsonString();
+      state.txApproval.asReceiver = asReceiver.toJsonString();
+      state.txApproval.transactionClassification = classifyTransaction(transactionView);
+
+      state.txApproval.attitude = undefined;
+    });
+  },
+
+  setAttitude: attitude => {
+    set(state => {
+      state.txApproval.attitude = attitude;
+    });
+  },
+
+  sendResponse: () => {
+    const {
+      responder,
+      attitude,
+      transactionView: transactionViewString,
+      authorizeRequest: authorizeRequestString,
+    } = get().txApproval;
+    try {
+      if (!responder || attitude === undefined || !transactionViewString || !authorizeRequestString)
+        throw new Error('Missing response data');
+
+      responder({
+        type: 'TX-APPROVAL',
+        data: {
+          attitude,
+          transactionView: TransactionView.fromJsonString(
+            transactionViewString,
+          ).toJson() as JsonObject,
+          authorizeRequest: AuthorizeRequest.fromJsonString(
+            authorizeRequestString,
+          ).toJson() as JsonObject,
+        },
+      });
+    } catch (err) {
+      if (responder)
+        responder({
+          type: 'TX-APPROVAL',
+          error: ConnectError.from(err),
+        });
+      else throw err;
+    } finally {
+      set(state => {
+        state.txApproval.responder = undefined;
+        state.txApproval.authorizeRequest = undefined;
+        state.txApproval.transactionView = undefined;
+        state.txApproval.attitude = undefined;
+      });
+    }
+  },
+});
 
 export const txApprovalSelector = (state: AllSlices) => state.txApproval;
+
+/**
+ * Returns a boolean indicating whether the given address is controlled by the
+ * current user.
```

Comment:
it looks like this is only ever used as a parameter for asRecieverTransactionView and its associated functions, and never anywhere else. so should probably be commented there and perhaps renamed to be less likely to confuse. to avoid confusion i've now tightened the scope and am no longer exporting it.

we might reconsider making view generation the responsibility of the component instead of the slice, which would allow us to use hooks/fetchers/queriers and permit this slice to become fully synchronous.